eth_block.go

```diff
@@ -0,0 +1,363 @@
+		Header:  &obj.Result.Header,
+		cid:     rawdataToCid(MEthBlock, headerRawData),
+		rawdata: headerRawData,
+	}
+
+	// Process the found eth-tx objects
+	ethTxNodes, ethTxTrieNodes := processTransactions(obj.Result.Transactions,
+		obj.Result.Header.TxHash[:])
+
+	return ethBlock, ethTxNodes, ethTxTrieNodes, nil
+}
+
+// processTransactions will take the found transactions in a parsed block body
+// to return IPLD node slices for eth-tx and eth-tx-trie
+func processTransactions(txs []*types.Transaction, expectedTxRoot []byte) ([]*EthTx, []*EthTxTrie) {
+	var ethTxNodes []*EthTx
+	transactionTrie := newTxTrie()
+
+	for idx, tx := range txs {
+		ethTx := NewTx(tx) // Will panic if it finds an error while parsing a tx
+		ethTxNodes = append(ethTxNodes, ethTx)
+		transactionTrie.addTx(idx, ethTx.RawData())
+	}
+
+	ethTxTrieNodes := transactionTrie.getNodes()
+	if !bytes.Equal(transactionTrie.rootHash(), expectedTxRoot) {
+		panic("Wrong transaction hash computed!")
+	}
+
+	return ethTxNodes, ethTxTrieNodes
+}
+
+/*
+  OUTPUT
+*/
+
+// DecodeEthBlock takes a cid and its raw binary data
+// from IPFS and returns an EthBlock object for further processing.
+func DecodeEthBlock(c *cid.Cid, b []byte) (*EthBlock, error) {
+	var h types.Header
+	err := rlp.Decode(bytes.NewReader(b), &h)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EthBlock{
+		Header:  &h,
+		cid:     c,
+		rawdata: b,
+	}, nil
+}
+
+/*
+  Block INTERFACE
+*/
+
+// RawData returns the binary of the RLP encode of the block header.
+func (b *EthBlock) RawData() []byte {
+	return b.rawdata
+}
+
+// Cid returns the cid of the block header.
+func (b *EthBlock) Cid() *cid.Cid {
+	return b.cid
+}
+
+// String is a helper for output
+func (b *EthBlock) String() string {
+	return fmt.Sprintf("<EthBlock %s>", b.cid)
+}
+
+// Loggable returns a map the type of IPLD Link.
+func (b *EthBlock) Loggable() map[string]interface{} {
+	return map[string]interface{}{
+		"type": "eth-block",
+	}
+}
+
+/*
+  Node INTERFACE
+*/
+
+// Resolve resolves a path through this node, stopping at any link boundary
+// and returning the object found as well as the remaining path to traverse
+func (b *EthBlock) Resolve(p []string) (interface{}, []string, error) {
+	if len(p) == 0 {
+		return b, nil, nil
+	}
+
+	switch p[0] {
+	case "bloom":
+		return b.Bloom, nil, nil // should not send the rest: You reached an end here.
+	case "coinbase":
+		return b.Coinbase, nil, nil
+	case "difficulty":
+		return b.Difficulty, nil, nil
+	case "extra":
+		return fmt.Sprintf("0x%x", b.Extra), nil, nil // This is a []byte. By default they are marshalled into Base64.
+	case "gaslimit":
+		return b.GasLimit, nil, nil
+	case "gasused":
+		return b.GasUsed, nil, nil
+	case "mixdigest":
+		return b.MixDigest, nil, nil
+	case "nonce":
+		return b.Nonce, nil, nil
+	case "number":
+		return b.Number, nil, nil
+	case "parent":
+		return &node.Link{Cid: commonHashToCid(MEthBlock, b.ParentHash)}, p[1:], nil
+	case "parentHash":
```

Comment:
Personally, I hate redundant information because it can easily get int an illegal state. Also, this means that anyone *constructing* a block using IPLD needs to specify these things twice. IMO, when working in IPLD land, programmers should convert all ETH hashes to/from IPLD CIDs at the program boundaries and use CIDs inside as much as possible.